Updated core files

JasonStoltz opened the pull request:
Note for reviewers: Reviewing the individual commits will provide a cleaner diff, since in the main diff some files are shown as new, rather than moved.

After implementing search on Elastic.co, there were a number of things that I added to the "core" files, that would be good to add here.

The "core" files here are a set of files which act as a "library" which will eventually be abstracted out into the UI Toolkit. The Reference UI should act as the source of truth for that code until then. Since search on Elastic.co essentially used a copy of those files from this project, it would be ideal to commit those changes back to the Reference UI.

Some of this behavior I'm committing back is not actually used in the Reference UI. However, if someone were to use this application as a starting point, it would benefit them to have that additional behavior available for use.

Updates include:
- Updated version of swiftype-app-search-javascript to the latest
- Factored out interactions with the swiftype-app-search-javascript client into its own file, called "AppSearchAPIConnector".
- Support for range filters
- Updated filter syntax to always use arrays internally, for simplicity
- Added more inline code documentation, to better serve as a reference
- Renamed `/app-search` to `/search-lib` and created an index.js file to make it feel more like a library rather than additional app code.
- Added additional actions: `clearFilters` and `setFilter`
- Support for conditional facets and disjunctive facets

There will be slight functional differences in the application:
- API calls are now cached, so performance will be a bit faster
- User input will now have an immediate effect on the UI, and the result area will be updated when the API request finishes.

goodroot commented on src/app-search/AppSearchDriver.js:
```diff
@@ -96,9 +92,9 @@ function formatORFiltersAsAND(filters = []) {
- * There's this weird thing where facet values for dates come back as an integer
- * from the API, but the API expects them as parameters formatted as date
- * strings.
+ * Facet values for dates come back as Integer from the API. However, the API
```
😄

goodroot replied:
😄

goodroot commented on README.md:
```diff
@@ -92,13 +92,13 @@ Logically, the pieces of this application fit together like this:
     ( State manager )       ( Syncs state with URL )
   -------------------      --------------
-  | AppSearchDriver | <--> | URLManager |
+  | SearchDriver     | <--> | URLManager |
```
This change makes a ton of sense. 👍

goodroot commented on src/search-lib/SearchDriver.js:
```diff
@@ -172,7 +172,7 @@ function matchFilter(filter1, filter2) {
  *   Configuration for Facet filters to be used within this application. The
  *   syntax for Facet configuration follows the API syntax:
- *   https://swiftype.com/documentation/app-search/api/search/facets. In
+ *   https://swiftype.com/documentation/search-lib/api/search/facets. In
```
This was mistakenly edited during the replace all. :)

JasonStoltz replied:
Oh geez, good catch 🤦‍♂️

goodroot commented on src/search-lib/SearchDriver.js:
```diff
@@ -214,12 +214,12 @@ function matchFilter(filter1, filter2) {
  *   This is low level configuration which lets you configure
  *   the options used on the Search API endpoint, ex: `result_fields`.
- *   https://swiftype.com/documentation/app-search/api/search
+ *   https://swiftype.com/documentation/search-lib/api/search
```
As was this.

goodroot commented on src/app-search/AppSearchAPIConnector.js:
```diff
@@ -0,0 +1,65 @@
+   * engineName  - Engine to query, found in your App Search Dashboard
+   * hostIdentifier - Credential found in your App Search Dashboard
+   * searchKey - Credential found in your App Search Dashboard
+   * endpointBase - (optional) Overrides the base of the Swiftype API endpoint
```
Awesome.

JasonStoltz commented:
@goodroot Read for re-review

JasonStoltz commented:
After implementing search on Elastic.co, there were a number of things that I added to the "core" files, that would be good to add here.

The "core" files here are a set of files which act as a "library" which will eventually be abstracted out into the UI Toolkit. The Reference UI should act as the source of truth for that code until then. Since search on Elastic.co essentially used a copy of those files from this project, it would be ideal to commit those changes back to the Reference UI.

Some of this behavior I'm committing back is not actually used in the Reference UI. However, if someone were to use this application as a starting point, it would benefit them to have that additional behavior available for use.

Updates include:

- Updated version of swiftype-app-search-javascript to the latest
- Factored out interactions with the swiftype-app-search-javascript client into its own file, called "AppSearchAPIConnector".
- Support for range filters
- Updated filter syntax to always use arrays internally, for simplicity
- Added more inline code documentation, to better serve as a reference
- Renamed /app-search to /search-lib and created an index.js file to make it feel more like a library rather than additional app code.
- Added additional actions: clearFilters and setFilter
- Support for conditional facets and disjunctive facets

There will be slight functional differences in the application:

- API calls are now cached, so performance will be a bit faster
- User input will now have an immediate effect on the UI, and the result area will be updated when the API request finishes.
